test/test_raw_api.py

```diff
@@ -0,0 +1,94 @@
+#
+# License https://www.backblaze.com/using_b2_code.html
+#
+######################################################################
+
+from __future__ import print_function
+
+import six
+
+from b2.raw_api import B2RawApi
+from b2.b2http import B2Http
+from b2.exception import UnusableFileName
+from .test_base import TestBase
+
+if six.PY3:
+    unichr = chr
+
+# Unicode characters for testing filenames.  (0x0394 is a letter Delta.)
+TWO_BYTE_UNICHR = unichr(0x0394)
+CHAR_UNDER_32 = unichr(31)
+DEL_CHAR = unichr(127)
+
+
+class TestRawAPIFilenames(TestBase):
+    """Test that the filename checker passes conforming names and rejects those that don't."""
+
+    def _should_be_ok(self, filename):
+        """Call with test filenames that follow the filename rules.
+
+        :param filename: unicode (or str) that follows the rules"""
+        print(u"Filename \"{}\" should be OK".format(filename))
+        self.assertIsNone(self.raw_api.check_b2_filename(filename))
+
+    def _should_raise(self, filename, exception_message):
+        """Call with filenames that don't follow the rules (so the rule checker should raise).
+
+        :param filename: unicode (or str) that doesn't follow the rules
+        :param exception_message: regexp that matches the exception's detailed message
+        """
+        print(
+            u"Filename \"{}\" should raise UnusableFileName(\".*{}.*\")."
+            .format(filename, exception_message)
+        )
+        with self.assertRaisesRegexp(UnusableFileName, exception_message):
+            self.raw_api.check_b2_filename(filename)
+
+    def test_b2_filename_checker(self):
+        """Test a conforming and non-conforming filename for each rule.
+
+        From the B2 docs (https://www.backblaze.com/b2/docs/files.html):
+        - Names can be pretty much any UTF-8 string up to 1024 bytes long.
+        - No character codes below 32 are allowed.
+        - Backslashes are not allowed.
+        - DEL characters (127) are not allowed.
+        - File names cannot start with "/", end with "/", or contain "//".
+        - Maximum of 250 bytes of UTF-8 in each segment (part between slashes) of a file name.
+        """
+        print("test b2 filename rules")
+        self.raw_api = B2RawApi(B2Http())
+
+        # Examples from doc:
+        self._should_be_ok('Kitten Videos')
+        self._should_be_ok(u'\u81ea\u7531.txt')
+
+        # Check length
+        # 1024 bytes is ok if the segments are at most 250 chars.
+        s_1024 = 4 * (250 * 'x' + '/') + 20 * 'y'
+        self._should_be_ok(s_1024)
+        # 1025 is too long.
+        self._should_raise(s_1024 + u'x', "too long")
+        # 1024 bytes with two byte characters should also work.
+        s_1024_two_byte = 4 * (125 * TWO_BYTE_UNICHR + u'/') + 20 * u'y'
+        self._should_be_ok(s_1024_two_byte)
+        # But 1025 bytes is too long.
+        self._should_raise(s_1024_two_byte + u'x', "too long")
```

Comment:
Glad you tested this case.  Chars vs. bytes always cause confusion, and this makes sure it's counting bytes.